diff --git a/include/assimp/config.h.in b/include/assimp/config.h.in
index bf0076572..d8bb11f4d 100644
--- a/include/assimp/config.h.in
+++ b/include/assimp/config.h.in
@@ -1116,14 +1116,4 @@ enum aiComponent
 #   define AI_CONFIG_APP_SCALE_KEY 1.0
 #endif // AI_CONFIG_APP_SCALE_KEY
 
-
-// ---------- All the Build/Compile-time defines ------------
-
-/** @brief Specifies if double precision is supported inside assimp
- *
- * Property type: Bool. Default value: undefined.
- */
-
-#cmakedefine ASSIMP_DOUBLE_PRECISION 1
-
 #endif // !! AI_CONFIG_H_INC
