diff -ruN A/BUILD.bazel B/BUILD.bazel
--- A/BUILD.bazel	2024-02-07 15:37:47
+++ B/BUILD.bazel	2024-02-07 15:41:20
@@ -82,7 +82,7 @@
     outs = ["generators.inc"],
     cmd = "$(location generate_registry_tables) --xml=$(location @spirv_headers//:spirv_xml_registry) --generator-output=$(location generators.inc)",
     cmd_bat = "$(location //:generate_registry_tables) --xml=$(location @spirv_headers//:spirv_xml_registry) --generator-output=$(location generators.inc)",
-    exec_tools = [":generate_registry_tables"],
+    tools = [":generate_registry_tables"],
 )
 
 py_binary(
@@ -96,7 +96,7 @@
     outs = ["build-version.inc"],
     cmd = "SOURCE_DATE_EPOCH=0 $(location update_build_version) $(location CHANGES) $(location build-version.inc)",
     cmd_bat = "set SOURCE_DATE_EPOCH=0  && $(location //:update_build_version) $(location CHANGES) $(location build-version.inc)",
-    exec_tools = [":update_build_version"],
+    tools = [":update_build_version"],
 )
 
 # Libraries
diff -ruN A/build_defs.bzl B/build_defs.bzl
--- A/build_defs.bzl	2024-02-07 15:37:47
+++ B/build_defs.bzl	2024-02-07 15:41:37
@@ -76,7 +76,7 @@
             "--core-insts-output=$(location {3}) " +
             "--operand-kinds-output=$(location {4})"
         ).format(*fmtargs),
-        exec_tools = [":generate_grammar_tables"],
+        tools = [":generate_grammar_tables"],
         visibility = ["//visibility:private"],
     )
 
